halo2-wasm/src/halo2lib/ecc.rs

```diff
@@ -0,0 +1,414 @@
+        let fq_chip = self.bn254_fq_chip();
+        let g1_chip = EccChip::new(&fq_chip);
+        self.load_bn254_g1_impl(&g1_chip, point)
+    }
+    /// `g1_points` should be array of `CircuitBn254G1Affine` in hi-lo form.
+    /// This function does not range check `hi,lo` to be `uint128` in case it's already done elsewhere.
+    /// Prevents any g1_points from being identity.
+    pub fn bn254_g1_sum(&self, g1_points: js_sys::Array) -> Bn254G1AffinePoint {
+        let fq_chip = self.bn254_fq_chip();
+        let g1_chip = EccChip::new(&fq_chip);
+        let g1_points: Vec<JsCircuitBn254G1Affine> = g1_points
+            .iter()
+            .map(|x| serde_wasm_bindgen::from_value(x).unwrap())
+            .collect();
+        let mut builder = self.builder.borrow_mut();
+        let ctx = builder.main(0);
+        let g1_points: Vec<_> = g1_points
+            .into_iter()
+            .map(|point| self.load_bn254_g1_impl(&g1_chip, point).0)
+            .collect();
+        let sum = g1_chip.sum::<Bn254G1Affine>(ctx, g1_points);
+        Bn254G1AffinePoint(sum)
+    }
+    /// Doesn't range check limbs of g2_point.
+    /// Does not allow you to load identity point.
+    pub fn load_bn254_g2(&self, point: JsCircuitBn254G2Affine) -> Bn254G2AffinePoint {
+        let fq_chip = self.bn254_fq_chip();
+        let fq2_chip = Bn254Fq2Chip::new(&fq_chip);
+        let g2_chip = EccChip::new(&fq2_chip);
+        self.load_bn254_g2_impl(&g2_chip, point)
+    }
+    /// `g2_points` should be array of `CircuitBn254G2Affine` in hi-lo form.
+    /// This function does not range check `hi,lo` to be `uint128` in case it's already done elsewhere.
+    /// Prevents any g2_points from being identity.
+    pub fn bn254_g2_sum(&self, g2_points: js_sys::Array) -> Bn254G2AffinePoint {
+        let fq_chip = self.bn254_fq_chip();
+        let fq2_chip = Bn254Fq2Chip::new(&fq_chip);
+        let g2_chip = EccChip::new(&fq2_chip);
+        let g2_points: Vec<JsCircuitBn254G2Affine> = g2_points
+            .iter()
+            .map(|x| serde_wasm_bindgen::from_value(x).unwrap())
+            .collect();
+        let mut builder = self.builder.borrow_mut();
+        let ctx = builder.main(0);
+        let g2_points: Vec<_> = g2_points
+            .into_iter()
+            .map(|point| self.load_bn254_g2_impl(&g2_chip, point).0)
+            .collect();
+        let sum = g2_chip.sum::<Bn254G2Affine>(ctx, g2_points);
+        Bn254G2AffinePoint(sum)
+    }
+    /// Verifies that e(lhs_1, rhs_1) = e(lhs_2, rhs_2) by checking e(lhs_1, rhs_1)*e(-lhs_2, rhs_2) === 1
+    /// Returns [CircuitValue] for the result as a boolean (1 if signature verification is successful).
+    /// None of the points should be identity.
+    pub fn bn254_pairing_check(
+        &self,
+        lhs_1: Bn254G1AffinePoint,
+        rhs_1: Bn254G2AffinePoint,
+        lhs_2: Bn254G1AffinePoint,
+        rhs_2: Bn254G2AffinePoint,
```

Comment:
why not (lhs_1, lhs_2, rhs_1, rhs_2) ordering?

Reply:
I think it's a little more standard to group by the things you are pairing together - this is in line with https://eips.ethereum.org/EIPS/eip-197

Reply:
ahh I see misunderstanding on my part. I would suggest changing the "lhs/rhs" naming then, because I was reading those as being lhs and rhs of the equality check. Hence I was assuming that lhs_1 and lhs_2 were the g1 and g2 points of the first pairing (pairing on the lhs of the equality check), whereas iiuc you're using lhs/rhs to mean lhs/rhs of a single pairing and 1/2 to mean pairing 1/2?

Reply:
ahhh yes, I think my naming here is indeed bad, I will switch to `lhs_g1, lhs_g2` etc.